### PR TITLE
BUG: pd.show_versions: json.decoder.JSONDecodeError

### LINKED ISSUE #39701: BUG: pd.show_versions: json.decoder.JSONDecodeError
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import contextlib, io, json
import pandas as pd

redirect = io.StringIO()
with contextlib.redirect_stdout(redirect):
    pd.show_versions(as_json=True)  # produces invalid json
pd_versions = redirect.getvalue()
pd_versions = json.loads(pd_versions)  # raises JSONDecodeError
```

#### Problem description

`pd.show_versions(as_json=True)` is supposed to print valid JSON, but it really doesn't. It is supposed to work as advertised but it doesn't. It prints invalid JSON that is not parsable. The following error is printed when attempting to parse it:

> json.decoder.JSONDecodeError: Expecting property name enclosed in double quotes: line 1 column 2 (char 1)

This functionality is important because when running a job in the cloud, it is not feasible to print a multiline output. This is because every printed line has significant attached metadata. It is better to have valid JSON output instead in this case. The implementation of the `show_versions` function looks to be very hacky and poorly reviewed, given that it fails in this very basic way.

#### Expected Output

In the code example, `json.loads(pd_versions)` should have been able to parse the captured text.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 7d32926db8f7541c356066dcadabf854487738de
python           : 3.8.6.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 19.6.0
Version          : Darwin Kernel Version 19.6.0: Tue Nov 10 00:10:30 PST 2020; root:xnu-6153.141.10~1/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : en_US.UTF-8
pandas           : 1.2.2
numpy            : 1.19.4
pytz             : 2020.4
dateutil         : 2.8.1
pip              : 21.0.1
setuptools       : 50.3.2
Cython           : None
pytest           : 6.2.2
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 3.0.0
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by alexprincel**
- [ x] closes #39701
- [ x] tests added / passed
- [ x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/util/test_show_versions.py [hunk 1]
```diff
@@ -26,21 +33,44 @@
     "ignore:Distutils:UserWarning"
 )
 @pytest.mark.filterwarnings("ignore:Setuptools is replacing distutils:UserWarning")
-def test_show_versions(capsys):
+@pytest.mark.parametrize("as_json", [True, False, "test_output.json"], indirect=True)
```
Comment (simonjayhawkins) [inline pandas/tests/util/test_show_versions.py:36]: why do we need the indirect parameterisation of the fixture? There are no composite fixtures in play here?

probably better to parametrize the fixture directly, or just parameterise the test function and remove the fixture.

File: pandas/util/_print_versions.py [hunk 2]
```diff
@@ -107,7 +107,7 @@ def show_versions(as_json: Union[str, bool] = False) -> None:
         j = {"system": sys_info, "dependencies": deps}
 
         if as_json is True:
-            print(j)
+            print(json.dumps(j))
```
Comment (simonjayhawkins) [inline pandas/util/_print_versions.py:110]: when printing to file, `json.dump(j, f, indent=2)` is used. would it be better to have the console output and file output the same?

maybe change the test, to ensure this consistency.

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @alexprincel for the PR.

**Review by lithomas1 — COMMENTED** [review]
Can you add a note in the whatsnew?(```doc/source/whatsnew/v1.3.0.rst```, I think this should go under Other in the section bug fixes.)

File: pandas/tests/util/test_show_versions.py (hunk 1 above)
Comment (jreback) [reply pandas/tests/util/test_show_versions.py:36]: yeah this is only used once, just paramterize directly

File: pandas/tests/util/test_show_versions.py [hunk 3]
```diff
@@ -26,21 +33,44 @@
...
+
+    pd.show_versions(as_json=as_json)
     captured = capsys.readouterr()
     result = captured.out
 
-    # check header
-    assert "INSTALLED VERSIONS" in result
+    # check valid json is printed to the console if as_json is True
+    if as_json is True:
+        json.loads(result)
```
Comment (jreback) [inline pandas/tests/util/test_show_versions.py:48]: can you assert somethings here

**Review by jreback — CHANGES_REQUESTED** [review]
pls also add a whatsnew note. bug fixes in I/O sectoin are ok

**Comment by alexprincel** [comment]
Not sure if this needs commenting (again, first-timer), but I believe I have implemented all requested changes and made testing more robust.

**Comment by lithomas1** [comment]
@alexprincel Looking at the diff, it looks like a merge went wrong somewhere. Can you fix this up?

**Comment by alexprincel** [comment]
> @alexprincel Looking at the diff, it looks like a merge went wrong somewhere. Can you fix this up?

Sorry about that. What seems to be the problem? I will do my best to resolve it. My git skills are not yet up to par.

**Comment by lithomas1** [comment]
@alexprincel If you click on ```Files changed```, you can see that there are unrelated changes in the ```pandas/tests/io``` directory. (Which I suspect is the result of a bad merge) I'm not really an expert on git either, so maybe you can ask one of the other maintainers for help.(If necessary you can revert the changes manually)

**Comment by alexprincel** [comment]
> @alexprincel If you click on `Files changed`, you can see that there are unrelated changes in the `pandas/tests/io` directory. (Which I suspect is the result of a bad merge) I'm not really an expert on git either, so maybe you can ask one of the other maintainers for help.(If necessary you can revert the changes manually). Apart from the unrelated changes, LGTM.

I saw this as well after. I have been reading on git this week and trying some things on test repos just to understand exactly what I need to do to correct this. If I can't figure it out I'll just start a new branch, put the modified files in that branch's working directory and commit that.

**Comment by MarcoGorelli** [comment]
If you only need to revert a handful of files, I'd suggest
```bash
git remote -v  # check that `upstream` is pandas-dev/pandas and `origin` is alexprincel/pandas
git fetch upstream master
git checkout upstream/master -- pandas/tests/io
git commit -m 'revert unrelated changes'
git push origin alexdev
```

File: pandas/tests/util/test_show_versions.py [hunk 4]
```diff
@@ -26,21 +30,65 @@
     "ignore:Distutils:UserWarning"
 )
 @pytest.mark.filterwarnings("ignore:Setuptools is replacing distutils:UserWarning")
-def test_show_versions(capsys):
+@pytest.mark.parametrize("as_json", [True, False, "test_output.json"])
+def test_show_versions(capsys, as_json, tmpdir):
     # gh-32041
-    pd.show_versions()
+    if isinstance(as_json, str):
+        as_json = os.path.join(tmpdir, as_json)
```
Comment (MarcoGorelli) [inline pandas/tests/util/test_show_versions.py:37]: This test is hard to follow can you make separate tests? See [don't put logic in tests](https://testing.googleblog.com/2014/07/testing-on-toilet-dont-put-logic-in.html)

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: pandas/tests/util/test_show_versions.py (hunk 4 above)
Comment (alexprincel) [reply pandas/tests/util/test_show_versions.py:37]: Thank you for those comments! Will implement them as soon as I figure out why isort is failing here but not on my local environment..

Comment (MarcoGorelli) [reply pandas/tests/util/test_show_versions.py:37]: my guess is that you haven't fetched and merged latest upstream changes, there have been some recent changes

Comment (alexprincel) [reply pandas/tests/util/test_show_versions.py:37]: I have separated it into multiple smaller functions in my local branch now. I'm looking at all the warning silencing that is done and wondering whether this is still useful. Looking at the git blame, the original author added this saying it would remove some warnings during CI. I tested removing those warning silencing and running pre-commit on my end and I do not see any warnings. Similarly, other test files do not have this warning. Should I go ahead and remove the warning filtering ?

Comment (MarcoGorelli) [reply pandas/tests/util/test_show_versions.py:37]: @alexprincel which warnings are you referring to? pre-commit just runs linting/formatting jobs, it wouldn't pick up on warnings given by pandas

Comment (alexprincel) [reply pandas/tests/util/test_show_versions.py:37]: I'm referring to the multiple calls to decorator @pytest.mark.filterwarnings. I am still unfamiliar with some aspects of purest, amongst other things, but removing those decorators does not result in warnings being raised when running pytest on test_show_version.

**Review by lithomas1 — APPROVED** [review]

File: pandas/tests/util/test_show_versions.py (hunk 4 above)
Comment (MarcoGorelli) [reply pandas/tests/util/test_show_versions.py]: Ah OK, previously you'd written

> I tested removing those warning silencing and running pre-commit on my end

I guess you meant `pytest` - now this makes sense to me, thanks

File: pandas/tests/util/test_show_versions.py [hunk 5]
```diff
@@ -44,3 +94,15 @@ def test_show_versions(capsys):
...
+
+
+def test_json_output_match(capsys, tmpdir):
+    pd.show_versions(as_json=True)
+    result_console = capsys.readouterr().out
+
+    out_path = os.path.join(tmpdir, "test_json.json")
+    pd.show_versions(as_json=out_path)
+    with open(out_path) as out_fd:
+        result_file = "".join(out_fd.readlines())
```
Comment (MarcoGorelli) [inline pandas/tests/util/test_show_versions.py:106]: is this the same as `out_fd.read()`? If so, then maybe that's a bit simpler?

File: pandas/tests/util/test_show_versions.py [hunk 6]
```diff
@@ -26,13 +33,56 @@
...
+
+    with open(as_json) as fd:
+        contents = fd.readlines()
+        str_contents = "".join(contents)
+
+        # make sure that there was output to the file
+        assert str_contents
+
+        # check if file output is valid JSON, will raise an exception if not
+        dict_check = json.loads(str_contents)
```
Comment (MarcoGorelli) [inline pandas/tests/util/test_show_versions.py:52]: any reason to do this over
```python
    with open(as_json) as fd:
        dict_check = json.load(fd)
```
?

Also, I think the naming is a bit confusing, maybe rename `dict_check` to `result` and `version_elements` to `expected`

File: pandas/tests/util/test_show_versions.py [hunk 7]
```diff
@@ -26,13 +33,56 @@
...
 )
 @pytest.mark.filterwarnings("ignore:Setuptools is replacing distutils:UserWarning")
-def test_show_versions(capsys):
+def test_show_versions(tmpdir):
+    as_json = os.path.join(tmpdir, "test_output.json")
+
+    pd.show_versions(as_json=as_json)
+
+    # make sure that the file was created
+    assert os.path.exists(as_json)
```
Comment (MarcoGorelli) [inline pandas/tests/util/test_show_versions.py:42]: the next lines will fail if this isn't true, so this strikes me as redundant, I think it can be removed

File: pandas/tests/util/test_show_versions.py [hunk 8]
```diff
@@ -26,13 +33,56 @@
...
+
+
+def test_show_versions_console(capsys):
     # gh-32041
-    pd.show_versions()
+    pd.show_versions(as_json=False)
     captured = capsys.readouterr()
     result = captured.out
 
-    # check header
```
Comment (MarcoGorelli) [inline pandas/tests/util/test_show_versions.py:35]: why remove this comment?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: pandas/tests/util/test_show_versions.py (hunk 6 above)
Comment (alexprincel) [reply pandas/tests/util/test_show_versions.py:52]: None that I can think of. It seems I made this more complicated than it need to be. Thanks !

File: pandas/tests/util/test_show_versions.py [hunk 9]
```diff
@@ -26,11 +37,44 @@
...
+
+    assert result == expected
+
+
+def test_show_versions_console_json(capsys):
+    pd.show_versions(as_json=True)
+    result = get_captured_output(capsys)
+
+    # check valid json is printed to the console if as_json is True
+    result = json.loads(result)
```
Comment (MarcoGorelli) [inline pandas/tests/util/test_show_versions.py:63]: instead of reassigning to `result` (which may cause problems when we expand mypy checks on tests) could you just do something like
```
stdout = capsys.readouterr().out
result = json.loads(stdout)
```
I'm not sure that making a function for `get_captured_output` adds much, repetition is more acceptable in tests than it is in source code

File: pandas/tests/util/test_show_versions.py [hunk 10]
```diff
@@ -26,11 +37,44 @@
     "ignore:Distutils:UserWarning"
 )
 @pytest.mark.filterwarnings("ignore:Setuptools is replacing distutils:UserWarning")
-def test_show_versions(capsys):
+def test_show_versions(tmpdir):
```
Comment (MarcoGorelli) [inline pandas/tests/util/test_show_versions.py:40]: below every test, can you add a comment with the issue number? something like
```
# GH39701
```
, see other recent tests

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Thanks @alexprincel for updating - I think this almost ready, just a couple of minor comments

**Comment by alexprincel** [comment]
> Thanks @alexprincel for updating - I think this almost ready, just a couple of minor comments

Will make them later today. Thanks for your comments, helps me improve.

**Comment by alexprincel** [comment]
I have addressed all comments. However, three errors  keep coming up during CI and I can't figure out why. pandas-dev.pandas (+ Linux py37_np_dev + Linux py39) all fail but I do not understand the error messages at all.

File: pandas/tests/util/test_show_versions.py [hunk 11]
```diff
@@ -1,10 +1,21 @@
...
+from pandas.util._print_versions import (
+    _get_dependency_info,
+    _get_sys_info,
+)
+
 import pandas as pd
 
 
+def get_captured_output(capsys):
+    return capsys.readouterr().out
```
Comment (MarcoGorelli) [inline pandas/tests/util/test_show_versions.py:16]: This is unused, can you remove it please?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Thanks @alexprincel - just one unused function to remove, other than that this looks good to me

File: pandas/tests/util/test_show_versions.py [hunk 12]
```diff
@@ -26,11 +33,47 @@
...
+
+    assert result == expected
+
+
+def test_show_versions_console(capsys):
     # gh-32041
-    pd.show_versions()
-    captured = capsys.readouterr()
-    result = captured.out
+    # GH39701
```
Comment (MarcoGorelli) [inline pandas/tests/util/test_show_versions.py:74]: I should've picked up on this earlier, but this one didn't need changing, as it was in the existing test
```suggestion
    # gh-32041
```

**Review by MarcoGorelli — APPROVED** [review]
Pending green, this looks good to me

@simonjayhawkins any further comments, or good to merge?

**Comment by simonjayhawkins** [comment]
LGTM

**Comment by MarcoGorelli** [comment]
cool - merging then, as commit `02e6bc7` was green and the commit following that just changed a comment

Thank you @alexprincel !

**Comment by alexprincel** [comment]
> cool - merging then, as commit `02e6bc7` was green and the commit following that just changed a comment
> 
> Thank you @alexprincel !

Thank you for your patience on this! Hopefully I will be more efficient on my next PR.
